fl4health/clients/ensemble_client.py

```diff
@@ -0,0 +1,149 @@
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        """
+        This client enables the training of ensemble models in a federated manner.
+
+        Args:
+            data_path (Path): path to the data to be used to load the data for client-side training
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
+                checkpointing. Defaults to None.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+
+        self.model: EnsembleModel
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        Then set initialized attribute to True. Also perform some checks to ensure the keys of the
+        optimizer dictionary are consistent with the model keys.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        super().setup_client(config)
+
+        assert len(self.optimizers) == len(self.model.ensemble_models)
+        assert all(
+            opt_key == model_key
+            for opt_key, model_key in zip(sorted(self.optimizers.keys()), sorted(self.model.ensemble_models.keys()))
+        )
+
+    def set_optimizer(self, config: Config) -> None:
+        """
+        Method called in the the setup_client method to set optimizer attribute returned by used-defined get_optimizer.
+        Ensures that the return value of get_optimizer is a dictionary since that is required for the ensemble client.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        optimizers = self.get_optimizer(config)
+        assert isinstance(optimizers, dict)
+        self.optimizers = optimizers
+
+    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> Tuple[Losses, Dict[str, torch.Tensor]]:
+        """
+        Given a single batch of input and target data, generate predictions
+        (both individual models and ensemble prediction), compute loss, update parameters and
+        optionally update metrics if they exist. (ie backprop on a single batch of data).
+        Assumes self.model is in train model already. Differs from parent method in that, there are multiple losses
```

Comment:
"Assumes self.model is in train model already" to "Assumes self.model is in train mode already"